src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,205 @@
+/**
+ * @file cne_impl.hpp
+ * @author Marcus Edel
+ * @author Kartik Nighania
+ *
+ * Conventional Neural Evolution
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_CNE_CNE_IMPL_HPP
+
+#include "cne.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+CNE::CNE(const size_t populationSize,
+         const size_t maxGenerations,
+         const double mutationProb,
+         const double mutationSize,
+         const double selectPercent,
+         const double finalValue,
+         const double fitnessHist) :
+         populationSize(populationSize),
+         maxGenerations(maxGenerations),
+         mutationProb(mutationProb),
+         mutationSize(mutationSize),
+         selectPercent(selectPercent),
+         finalValue(finalValue),
+         fitnessHist(fitnessHist)
+         { /* Nothing to do here. */ }
+
+//! Optimize the function
+template<typename DecomposableFunctionType>
+double CNE::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& answer)
+{
+  // Get the number of functions to iterate
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1]
+  population.set_size(populationSize, answer.n_rows);
+  population.randu();
+
+  // initialize helper variables
+  arma::vec parameters(answer.n_rows);
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+  << std::endl;
+
+  // Iterate till max number of generations
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // calculate fitness values of all candidates
+    for (size_t i = 0; i < populationSize; i++)
+     {
+       // select a candidate
+       parameters = population.row(i).t();
```

Comment:
It looks like we could just use ``answer`` here instead to go with ``parameters``, that would avoid an extra copy. What do you think?